Initialize guidance for dev containers (#381)
microsoft/code-with-engineering-playbook: developer-experience/devcontainers/readme.md

# Getting Started with Dev Containers

> If you are a developer and experienced Visual Studio Code (VS Code) or Docker, then it's probably time you look at [development containers](http://web.archive.org/web/20200806073925/https://code.visualstudio.com/docs/remote/containers) (dev containers). This readme is intended to assist developers in the decision-making process needed to build dev containers. The guidance provided should be especially helpful if you are experiencing VS Code dev containers for the first time.

Reviewer: nit: experienced Visual Studio Code => have experience with Visual Studio Code

> If you are a developer and experienced Visual Studio Code (VS Code) or Docker, then it's probably time you look at [development containers](http://web.archive.org/web/20200806073925/https://code.visualstudio.com/docs/remote/containers) (dev containers). This readme is intended to assist developers in the decision-making process needed to build dev containers. The guidance provided should be especially helpful if you are experiencing VS Code dev containers for the first time.

Reviewer: ```suggestion
If you are a developer and have experience with Visual Studio Code (VS Code) or Docker, then it's probably time you look at [development containers](http://web.archive.org/web/20200806073925/https://code.visualstudio.com/docs/remote/containers) (dev containers). This readme is intended to assist developers in the decision-making process needed to build dev containers. The guidance provided should be especially helpful if you are experiencing VS Code dev containers for the first time.
```

> Let's first introduce what dev containers are, briefly describe how they improve collaboration between developers and then highlight some high-level steps that you will be facing when building a dev container.

Reviewer: nit: this sentence could be removed and the message would still be the same - if we want to simplify the document

Author: Thnks. Intro has been shortened.

![image](https://user-images.githubusercontent.com/10041279/93239062-e1b9a480-f747-11ea-94fb-3d50b14fd9b1.png)

If the above diagram is not clear, a basic analogy that might help you intuitively understand dev containers is to think of them as a union between Docker’s interactive mode (`docker exec -it 987654e0ff32`) and the VS Code UI experience that you are used to.

To set yourself up for the dev container experience described above, use your VS Code’s Extension Marketplace to install the [Remote Development Extension Pack](http://web.archive.org/web/20200828150357/https://marketplace.visualstudio.com/items?itemName=ms-vscode-remote.vscode-remote-extensionpack).

## How can dev containers improve project collaboration?

VS Code dev containers have improved project collaboration between developers on recent team projects by addressing two very specific problems:

- Inconsistent local developer experiences within a team.

> 

Reviewer: nit: remove the empty line

> 

Reviewer: ```suggestion
```

- Slow onboarding of developers joining a project.

> The problems listed above were addressed by configuring and then sharing a dev container definition. Dev containers are defined by their base image and the artifacts that support that base image. The base image and the artifacts that come with it live in the .devcontainer directory. This directory is where configuration begins. A central artifact to the dev container definition is a json configuration file called `devcontainer`. This file orchestrates the artifacts needed to support the base image and the dev container lifecycle. Installation of the [Remote Development Extension Pack](http://web.archive.org/web/20200828150357/https://marketplace.visualstudio.com/items?itemName=ms-vscode-remote.vscode-remote-extensionpack) is required to enable this orchestration within a project repo.

Reviewer 2: ```suggestion
The problems listed above were addressed by configuring and then sharing a dev container definition. Dev containers are defined by their base image and the artifacts that support that base image. The base image and the artifacts that come with it live in the .devcontainer directory. This directory is where configuration begins. A central artifact to the dev container definition is a configuration file called `devcontainer.json`. This file orchestrates the artifacts needed to support the base image and the dev container lifecycle. Installation of the [Remote Development Extension Pack](http://web.archive.org/web/20200828150357/https://marketplace.visualstudio.com/items?itemName=ms-vscode-remote.vscode-remote-extensionpack) is required to enable this orchestration within a project repo.
```

        "args": {"VARIANT": "3.8"}
    },
}
```

For a list of devcontainer.json configuration properties, visit VS Code documentation on [dev container properties](https://code.visualstudio.com/docs/remote/devcontainerjson-reference).

## How do I decide which dev container is right for my use case?

Fortunately, VS Code has a repo gallery of platform specific folders that host dev container definitions (.devcontainer directories) to make getting started with dev containers easier. The code snippet below shows a list of gallery folders that come directly from the [VS Code dev container gallery repo](https://github.com/microsoft/vscode-dev-containers/tree/master/containers):

```bash
$ tree vs-code-dev-containers  # main repo directory
└───containers
        ├───dotnetcore
        |   └───.devcontainers # dev container
        ├───python-3
        |   └───.devcontainers # dev container
        ├───ubuntu
        |   └───.devcontainers # dev container
        └───....
```

Here are the final high-level steps it takes to build a dev container:

1. Decide which platform you'd like to build a local development tool stack around.

> 

Reviewer: nit: remove empty lines in the list

> 

Reviewer: ```suggestion
```

2. Browse the VS Code provided dev container gallery of project folders that target your platform and choose the most appropriate one.

> 

Reviewer: ```suggestion
```

3. Inspect the dev container definitions (.devcontainer directory) of a project for the base image and the artifacts that support that base image.

> 

Reviewer: ```suggestion
```
